drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,751 @@
+ */
+static int eeprom_emu_compactor(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	off_t next_page_offset;
+	int rc;
+
+	LOG_DBG("Compactor called for page at [0x%tx]",
+		(ptrdiff_t)dev_data->page_offset);
+
+	next_page_offset = dev_data->page_offset + dev_config->page_size;
+	if (next_page_offset >= dev_config->flash_size) {
+		next_page_offset = 0;
+	}
+
+	/* erase the new page */
+	rc = eeprom_emu_flash_erase(dev, next_page_offset,
+				    dev_config->page_size);
+	if (rc) {
+		return rc;
+	}
+
+	if (dev_config->rambuf) {
+		rc = eeprom_emu_flash_write(dev, next_page_offset,
+					    dev_config->rambuf,
+					    dev_config->size);
+		if (rc) {
+			return rc;
+		}
+
+	} else {
+		off_t rd_offset = 0;
+		uint8_t buf[dev_config->flash_cbs];
+
+		/* copy existing data */
+		while (rd_offset < dev_config->size) {
+			rc = eeprom_emu_word_read(dev, rd_offset, buf);
+			if (rc) {
+				return rc;
+			}
+
+			if (eeprom_emu_is_word_used(dev, buf)) {
+				rc = eeprom_emu_flash_write(dev,
+							    next_page_offset +
+							    rd_offset, buf,
+							    sizeof(buf));
+				if (rc) {
+					return rc;
+				}
+
+			}
+
+			rd_offset += sizeof(buf);
+		}
+
+	}
+
+	/* invalidate the old page */
+	rc = eeprom_emu_page_invalidate(dev, dev_data->page_offset);
+	if (!rc) {
+		dev_data->write_offset = dev_config->size;
+		dev_data->page_offset = next_page_offset;
+	}
+
+	return rc;
+}
+
+
+/*
+ * eeprom_emu_word_write: write basic word (cbs bytes of data) item to address,
+ */
+static int eeprom_emu_word_write(const struct device *dev, off_t address,
+				 const uint8_t *data)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	uint8_t buf[dev_config->flash_cbs], tmp[dev_config->flash_cbs];
+	off_t direct_address, wraddr;
+	int rc;
+
+	direct_address = dev_data->page_offset + address;
+
+	rc = eeprom_emu_flash_read(dev, direct_address, buf, sizeof(buf));
+	if (rc) {
+		return rc;
+	}
+
+	if (!eeprom_emu_is_word_used(dev, buf)) {
+		if (eeprom_emu_is_word_used(dev, data)) {
+			rc = eeprom_emu_flash_write(dev, direct_address, data,
+						    sizeof(buf));
+		}
+
+		goto end;
+	}
+
+	rc = eeprom_emu_word_read(dev, address, buf);
+	if (rc) {
+		return rc;
+	}
+
+	if (!memcmp(buf, data, sizeof(buf))) {
+		/* data has not changed */
+		goto end;
+	}
+
+	wraddr = address;
+	/* store change */
+	for (uint8_t i = 0; i < 2; i++) {
+		if (memcmp(&buf[i*sizeof(buf)/2], data, sizeof(buf)/2)) {
+			eeprom_emu_set_change(dev, wraddr, data, tmp);
+			rc = eeprom_emu_flash_write(dev, dev_data->page_offset +
+						    dev_data->write_offset, tmp,
+						    sizeof(buf));
+			if (rc) {
+				return rc;
+			}
+
+			dev_data->write_offset += sizeof(buf);
+			if ((dev_data->write_offset + sizeof(buf)) >=
+			    dev_config->page_size) {
+				rc = eeprom_emu_compactor(dev);
```

Comment:
It would be possible to detect GC need before any change record will be written.
Thanks to that refreshed `EEPROM data` entry will contain change at once.

Reply:
It is not so easy to detect GC need,  a write can be done to a already used part and an unused part. However i do understand what you mean. I will give it some thinking and try to come up with a prposal (the proposal has been postponed to 2.6 anyhow).

Reply:
The lastest commit should resolve this.